sharktank/sharktank/types/tensors.py

```diff
@@ -1157,22 +1186,55 @@ def create(
             if "shard_dim" in extra_properties
             else None
         )
-        ts = []
-        for i in range(shard_count):
-            t_name = str(i)
-            try:
-                t = raw_tensors[t_name]
-                ts.append(t)
-                # TODO: this should be changed to tracked device affinity
-                DeviceTensorTrait(i).set(t)
-            except KeyError as e:
-                raise IOError(
-                    f"Missing component tensor '{t_name}' in {raw_tensors.keys()}"
-                ) from e
+        shard_type = extra_properties.get(
+            "shard_type", DefaultPrimitiveTensor.serialized_name()
+        )
+        extra_properties_shards = extra_properties.get("extra_properties_shards", {})
+        try:
+            shard_clazz = REGISTERED_INFERENCE_TENSOR_CLASSES[shard_type]
+        except KeyError:
+            raise IOError(
+                f"Cannot deserialize ReplicatedTensor because of unregistered shard type"
+                f"{shard_type}"
+            )
+
+        raw_tensors = raw_tensors.copy()
+        raw_tensors_for_rank = {i: {} for i in range(shard_count)}
+        for k, v in raw_tensors.items():
+            if ".shard." in k:
+                # This is a shard tensor.
+                tensor_local_name, rank = k.split(".shard.")
+                rank = int(rank)
+                if rank < 0 or rank >= shard_count:
+                    raise IOError(
+                        f"Invalid shard rank {rank} for raw tensor {k} in {raw_tensors.keys()}."
+                    )
+                raw_tensors_for_rank[rank][tensor_local_name] = v
+            else:
+                import warnings
+
+                warnings.warn(
+                    "Using an old .irpa which relies of deprecated features that will be removed in a future update, please re-export weights to .irpa.",
+                    DeprecationWarning,
+                    stacklevel=2,
+                )
+                raw_tensors_for_rank[int(k)][""] = v
+        shards = []
+        for i, raw_tensors in raw_tensors_for_rank.items():
+            shard = shard_clazz.create(
+                name=f"{name}.shard.{i}",
+                raw_tensors=raw_tensors,
+                extra_properties=extra_properties_shards.get(str(i), {}),
```

Comment:
It seems this together with the `shard_type` default would give us backward compatibility to be able to load old archives.

Reply:
Yes, I was going for that. All of the current tests that use existing IRPAs (e.g. the sharded perplexity tests run without issues).